Update bdr mode && add description related bdr sql (#15865)
pingcap/docs-cn: ticdc/ticdc-bidirectional-replication.md

这样，以上搭建好的集群即可对数据进行双向复制。

## DDL 同步

> 从 v7.6.0 版本之后，TiDB 为了支持双向同步 DDL，引入了四种 BDR role（None、Primary、Secondary 和 Local_only）和两种 BDR 模式。

Reviewer: ```suggestion
从 v7.6.0 版本之后，TiDB 为了支持双向同步 DDL，引入了 `NONE`、`PRIMARY`、`SECONDAR` 和 `LOCAL_ONLY` 四种 BDR role 和两种 BDR 模式。
```

Reviewer 2: 严格和非严格表达上面是不是不太合适，自动和手动 BDR 模式是不是好些？

### BDR 严格模式

> 把一个 TiDB 集群设置为 Primary，其他 TiDB 集群设置为 Secondary 就可以进入 BDR 严格模式。在这种模式下
> 1. 用户可以在 primary TiDB 下执行无损 DDL，并通过 TiCDC 同步到 Secondary 集群。无损 DDL 如下：

Reviewer: ```suggestion
把一个 TiDB 集群设置为 `PRIMARY`，其他 TiDB 集群设置为 `SECONDARY`，就可以进入 BDR 严格模式。在这种模式下，你可以在 `PRIMARY` TiDB 下执行无损 DDL，并通过 TiCDC 同步到 `SECONDARY` 集群。无损 DDL 如下：
```

> 1. 用户可以在 primary TiDB 下执行无损 DDL，并通过 TiCDC 同步到 Secondary 集群。无损 DDL 如下：

Reviewer 2: 有损 DDL modify column 不能自动同步么？

> 1. 用户可以在 primary TiDB 下执行无损 DDL，并通过 TiCDC 同步到 Secondary 集群。无损 DDL 如下：

Reviewer 2: 我们无损/有损 DDL 好像不是通用术语哦。

Author: 稍等约个时间我们一起定一下

- create database
- create table
- add column(not null or default value)
- add non-unique index
- drop index

> - modify column(can modify default value and comment)

Reviewer 3: ```suggestion
- modify column(Only modifications to the default value and comment are permitted)
```

- alter column default value
- modify table comment
- rename index
- add table partition
- drop primary key
- alter table index visibility
- alter table ttl
- alter table remove ttl
2. 其他有损 DDL 在 primary TiDB 上执行会被拦截。有损 DDL 如下：
- drop database
- drop table
- add column 非（not null or default value）
- drop column
- add unique index
- truncate table
- modify column 非（can modify default value and comment）
- rename table
- drop partition
- truncate partition
- alter table character set
- alter database character set
- recover table
- add primary key
- rebase auto id
- exchange partition

> - reorganize partition

Reviewer: ```suggestion
- reorganize partition

```

> 3. 无论是有损 DDL 还是无损 DDL 都无法在 Secondary TiDB 中执行。在 TiCDC zhi

Reviewer 3: TiCDC zhi  这里是不是还没写完？ @okJiang

> 都可以执行的 DDL

Reviewer: ```suggestion
都可以执行的 DDL 有：

```

create view

> drop view

Reviewer 3: 既然都可以执行，是不是纳入 safe DDL ？然后说 secondary Cluster 除了 safe DDL 中的create view 和 drop view 其他 DDL 都无法被执行。

Author: 这个还要和 zongqi 商量，等他有空

这样，以上搭建好的集群即可对数据进行双向复制。

## DDL 类别

从 v7.6.0 版本之后，为了在双向同步中尽可能地支持 DDL 同步，根据 DDL 对业务的影响，TiDB 把 [TiCDC 原本支持同步的 DDL](/ticdc/ticdc-ddl.md) 划分为了两种 DDL：可复制的 DDL 和不可复制的 DDL。

### 可复制的 DDL

可复制的 DDL 是指在双向同步中，可以直接执行并同步到其他 TiDB 集群的 DDL。

> 可复制的 DDL包括：

Reviewer 4: ```suggestion
可复制的 DDL 包括：
```

这样，以上搭建好的集群即可对数据进行双向复制。

## DDL 类别

从 v7.6.0 开始，为了在双向复制中尽可能地支持 DDL 同步，根据 DDL 对业务的影响，TiDB 将 [TiCDC 原本支持同步的 DDL](/ticdc/ticdc-ddl.md) 划分为两种 DDL：可复制的 DDL 和不可复制的 DDL。

### 可复制的 DDL

> 可复制的 DDL 是指在双向同步中，可以直接执行并同步到其他 TiDB 集群的 DDL。

Reviewer: ```suggestion
可复制的 DDL 是指在双向复制中，可以直接执行并同步到其他 TiDB 集群的 DDL。
```

可复制的 DDL 包括：

- `CREATE DATABASE`
- `CREATE TABLE`
- `ADD COLUMN`：添加的列必须是 `not null` 或者带有 `default value` 的列
- `ADD NON-UNIQUE INDEX`
- `DROP INDEX`
- `MODIFY COLUMN`：仅能修改列的 `default value` 和 `comment`
- `ALTER COLUMN DEFAULT VALUE`
- `MODIFY TABLE COMMENT`
- `RENAME INDEX`
- `ADD TABLE PARTITION`
- `DROP PRIMARY KEY`
- `ALTER TABLE INDEX VISIBILITY`
- `ALTER TABLE TTL`
- `ALTER TABLE REMOVE TTL`
- `CREATE VIEW`
- `DROP VIEW`

### 不可复制的 DDL

> 不可复制的 DDL 是指对业务影响较大、不能在双向同步中直接通过 TiCDC 同步到其他 TiDB 集群的 DDL。不可复制的 DDL 必须通过特定的操作来执行。

Reviewer: ```suggestion
不可复制的 DDL 是指对业务影响较大、不能在双向复制中直接通过 TiCDC 同步到其他 TiDB 集群的 DDL。不可复制的 DDL 必须通过特定的操作来执行。
```

- `ADD COLUMN`：添加的列为 `null` 且不带有 `default value` 的列
- `DROP COLUMN`
- `ADD UNIQUE INDEX`
- `TRUNCATE TABLE`
- `MODIFY COLUMN`：修改列除 `default value` 和 `comment` 以外的属性
- `RENAME TABLE`
- `DROP PARTITION`
- `TRUNCATE PARTITION`
- `ALTER TABLE CHARACTER SET`
- `ALTER DATABASE CHARACTER SET`
- `RECOVER TABLE`
- `ADD PRIMARY KEY`
- `REBASE AUTO ID`
- `EXCHANGE PARTITION`
- `REORGANIZE PARTITION`

## DDL 同步

为了能够解决上述可复制的 DDL 和不可复制的 DDL 两类 DDL 的同步问题，TiDB 引入了三种 BDR role：

- `LOCAL_ONLY`（默认）：你可以执行任意 DDL，但是在 TiCDC 开启 `bdr_mode=true` 之后，执行的 DDL 不会被 TiCDC 同步。
- `PRIMARY`：你可以执行可复制的 DDL，但不能执行不可复制的 DDL，可复制的 DDL 会被 TiCDC 同步到下游。
- `SECONDARY`：你不能执行可复制的 DDL，也不能执行不可复制的 DDL，但是会执行从 TiCDC 同步过来的 DDL。

> **警告：**
>
> 双向复制的 DDL 同步目前为实验特性，不建议在生产环境中使用。该功能可能会在未事先通知的情况下发生变化或删除。如果发现 bug，请在 GitHub 上提 [issue](https://github.com/pingcap/tidb/issues) 反馈。

### 可复制的 DDL 的同步场景

1. 选择一个 TiDB 集群，执行 `ADMIN SET BDR ROLE PRIMARY` 将其设置为主集群。
2. 在其他 TiDB 集群上，执行 `ADMIN SET BDR ROLE SECONDARY` 将其设置为从集群。
3. 在主集群上执行**可复制的 DDL**，执行成功的 DDL 会被 TiCDC 同步到从集群中。

> **注意：**
>
> 为了防止误操作：
>
> - 如果在主集群中尝试执行**不可复制的 DDL**，会[报错 8263](/error-codes.md)。
> - 无论在从集群中尝试执行**可复制的 DDL** 还是**不可复制的 DDL**，都会[报错 8263](/error-codes.md)。

### 不可复制的 DDL 的同步场景

> 1. 将所有 TiDB 集群的 BDR role 设置为 `LOCAL_ONLY`（默认值），然后执行 `ADMIN SET BDR ROLE LOCAL_ONLY`。

Reviewer 5: 这里没太懂，“将所有 TiDB 集群的 BDR role 设置为 `LOCAL_ONLY`”是指在每个集群上执行 `ADMIN SET BDR ROLE LOCAL_ONLY` 吗？
这里说的“然后执行”是指在哪里执行？

Author: > 这里没太懂，“将所有 TiDB 集群的 BDR role 设置为 LOCAL_ONLY”是指在每个集群上执行 ADMIN SET BDR ROLE LOCAL_ONLY 吗？

是的

> 这里说的“然后执行”是指在哪里执行？

感觉「然后」可以去掉

> 1. 将所有 TiDB 集群的 BDR role 设置为 `LOCAL_ONLY`（默认值），然后执行 `ADMIN SET BDR ROLE LOCAL_ONLY`。

Reviewer: ```suggestion
1. 执行 `ADMIN SET BDR ROLE LOCAL_ONLY`，将所有 TiDB 集群的 BDR role 设置为 `LOCAL_ONLY`（默认值）。
```

2. 暂停所有集群中需要执行 DDL 的对应的表的写入操作。
3. 等待所有集群中对应表的所有写入已经同步到其他集群后，手动在每一个 TiDB 集群上单独执行所有的 DDL。
4. 等待 DDL 完成之后，重新恢复写入。
5. 按照[可复制的 DDL 的同步场景](#可复制的-ddl-的同步场景)的操作步骤，切换回可复制的 DDL 的同步场景。

## 停止双向复制

在业务数据停止写入之后，你可以在两个集群中都插入一行特殊的值，通过检查这两行特殊的值来确保数据达到了一致的状态。

检查完毕之后，停止同步任务，并把 BDR role 切换回 `LOCAL_ONLY` 即可停止双向复制。

## 使用限制

- BDR role 只能在以下两种场景中正常使用：

    - 1 个 `PRIMARY` 集群和 n 个 `SECONDARY` 集群（可复制的 DDL 的同步场景）
    - n 个 `LOCAL_ONLY` 集群（不可复制的 DDL 的同步场景）

    **注意，请勿将 BDR role 设置为其他情况，例如，同时设置了 `PRIMARY`、`SECONDARY`、`LOCAL_ONLY`。如果错误地设置了 BDR role，TiDB 无法保证数据正确性。**

> - 一般情况下，禁止在同步的表中使用 `AUTO_INCREMENT` 或 `AUTO_RANDOM` 键，以免产生数据冲突的问题。如果需要使用 `AUTO_INCREMENT` 或 `AUTO_RANDOM` 键，可以通过在不同的集群设置 `auto_increment_increment` 和 `auto_increment_offset` 来使得不同的集群都能够分配到不同的 primary key。例如：
>   - 假设有三台 TiDB（A、B、C）处于双向同步中，那么我们可以：
>   - 在 A 中设置 `auto_increment_increment=3`，`auto_increment_offset=2000`
>   - 在 B 中设置 `auto_increment_increment=3`，`auto_increment_offset=2001`
>   - 在 C 中设置 `auto_increment_increment=3`，`auto_increment_offset=2002`
>   - 这样的话，A、B、C 隐式分配到的 `AUTO_INCREMENT` ID 和 `AUTO_RANDOM` ID 就不会互相冲突
>   - 如果需要增加 BDR 模式的集群，则要临时暂停相关业务的写入，重新在所有集群上设置合适的 `auto_increment_increment` 和 `auto_increment_offset` 后，再开启相关业务。

Reviewer 5: ```suggestion
- 一般情况下，禁止在同步的表中使用 `AUTO_INCREMENT` 或 `AUTO_RANDOM` 键，以免产生数据冲突的问题。如果需要使用 `AUTO_INCREMENT` 或 `AUTO_RANDOM` 键，可以通过在不同的集群设置 `auto_increment_increment` 和 `auto_increment_offset` 来使得不同的集群都能够分配到不同的 primary key。假设有三台 TiDB（A、B、C）处于双向同步中，那么你可以采取如下设置：

    - 在 A 中设置 `auto_increment_increment=3`，`auto_increment_offset=2000`
    - 在 B 中设置 `auto_increment_increment=3`，`auto_increment_offset=2001`
    - 在 C 中设置 `auto_increment_increment=3`，`auto_increment_offset=2002`

  这样的话，A、B、C 隐式分配到的 `AUTO_INCREMENT` ID 和 `AUTO_RANDOM` ID 就不会互相冲突。如果需要增加 BDR 模式的集群，则要临时暂停相关业务的写入，重新在所有集群上设置合适的 `auto_increment_increment` 和 `auto_increment_offset` 后，再开启相关业务。
```

这样，以上搭建好的集群即可对数据进行双向复制。

## DDL 类别

从 v7.6.0 开始，为了在双向复制中尽可能地支持 DDL 同步，根据 DDL 对业务的影响，TiDB 将 [TiCDC 原本支持同步的 DDL](/ticdc/ticdc-ddl.md) 划分为两种 DDL：可复制的 DDL 和不可复制的 DDL。

### 可复制的 DDL

可复制的 DDL 是指在双向复制中，可以直接执行并同步到其他 TiDB 集群的 DDL。

可复制的 DDL 包括：

- `CREATE DATABASE`
- `CREATE TABLE`
- `ADD COLUMN`：添加的列必须是 `not null` 或者带有 `default value` 的列
- `ADD NON-UNIQUE INDEX`
- `DROP INDEX`
- `MODIFY COLUMN`：仅能修改列的 `default value` 和 `comment`
- `ALTER COLUMN DEFAULT VALUE`
- `MODIFY TABLE COMMENT`
- `RENAME INDEX`
- `ADD TABLE PARTITION`
- `DROP PRIMARY KEY`
- `ALTER TABLE INDEX VISIBILITY`
- `ALTER TABLE TTL`
- `ALTER TABLE REMOVE TTL`
- `CREATE VIEW`
- `DROP VIEW`

### 不可复制的 DDL

> 不可复制的 DDL 是指对业务影响较大、不能在双向复制中直接通过 TiCDC 同步到其他 TiDB 集群的 DDL。不可复制的 DDL 必须通过特定的操作来执行。

Reviewer 3: ```suggestion
不可复制的 DDL 是指对业务影响较大、可能会造成集群间数据不一致性的 DDL，这类 DDL 不能在双向复制中直接通过 TiCDC 同步到其他 TiDB 集群的 DDL。不可复制的 DDL 必须通过特定的操作来执行。
```

- `ALTER DATABASE CHARACTER SET`
- `RECOVER TABLE`
- `ADD PRIMARY KEY`
- `REBASE AUTO ID`
- `EXCHANGE PARTITION`
- `REORGANIZE PARTITION`

## DDL 同步

为了能够解决上述可复制的 DDL 和不可复制的 DDL 两类 DDL 的同步问题，TiDB 引入了两种 BDR role：

- `PRIMARY`：你可以执行可复制的 DDL，但不能执行不可复制的 DDL，可复制的 DDL 会被 TiCDC 同步到下游。
- `SECONDARY`：你不能执行可复制的 DDL，也不能执行不可复制的 DDL，但是会执行从 TiCDC 同步过来的 DDL。

在不设置 BDR role 时：你可以执行任意 DDL，但是在 TiCDC 开启 `bdr_mode=true` 之后，执行的 DDL 不会被 TiCDC 同步。

> **警告：**
>
> 双向复制的 DDL 同步目前为实验特性，不建议在生产环境中使用。该功能可能会在未事先通知的情况下发生变化或删除。如果发现 bug，请在 GitHub 上提 [issue](https://github.com/pingcap/tidb/issues) 反馈。

### 可复制的 DDL 的同步场景

1. 选择一个 TiDB 集群，执行 `ADMIN SET BDR ROLE PRIMARY` 将其设置为主集群。
2. 在其他 TiDB 集群上，执行 `ADMIN SET BDR ROLE SECONDARY` 将其设置为从集群。
3. 在主集群上执行**可复制的 DDL**，执行成功的 DDL 会被 TiCDC 同步到从集群中。

> **注意：**
>
> 为了防止误操作：
>
> - 如果在主集群中尝试执行**不可复制的 DDL**，会[报错 8263](/error-codes.md)。
> - 无论在从集群中尝试执行**可复制的 DDL** 还是**不可复制的 DDL**，都会[报错 8263](/error-codes.md)。

### 不可复制的 DDL 的同步场景

1. 对所有 TiDB 集群执行 `ADMIN UNSET BDR ROLE`，撤销集群的 BDR role。
2. 暂停所有集群中需要执行 DDL 的对应的表的写入操作。
3. 等待所有集群中对应表的所有写入已经同步到其他集群后，手动在每一个 TiDB 集群上单独执行所有的 DDL。
4. 等待 DDL 完成之后，重新恢复写入。
5. 按照[可复制的 DDL 的同步场景](#可复制的-ddl-的同步场景)的操作步骤，切换回可复制的 DDL 的同步场景。

> 

Author: ```suggestion

> **注意：**
>
> 在此期间，请勿执行其他任意的 DDL！

```

Reviewer 3: LGTM

## 停止双向复制

在业务数据停止写入之后，你可以在两个集群中都插入一行特殊的值，通过检查这两行特殊的值来确保数据达到了一致的状态。

检查完毕之后，停止同步任务，并对所有集群执行 `ADMIN UNSET BDR ROLE`。

## 使用限制

- BDR role 只能在以下两种场景中正常使用：

    - 1 个 `PRIMARY` 集群和 n 个 `SECONDARY` 集群（可复制的 DDL 的同步场景）

>     - n 个不设置 BDR role 的集群（不可复制的 DDL 的同步场景）

Reviewer 3: ```suggestion
    - n 个不设置 BDR role 的集群（用于在每个集群手动执行不可复制的 DDL 的同步场景）
```

>     **注意，请勿将 BDR role 设置为其他情况，例如，既存在集群设置了 `PRIMARY`、`SECONDARY`，又存在集群没有设置 BDR role。如果错误地设置了 BDR role，TiDB 无法保证数据正确性。**

Reviewer 3: ```suggestion
    **注意，请勿将 BDR role 设置为其他情况，例如，既存在集群设置了 `PRIMARY`、`SECONDARY`，又存在集群没有设置 BDR role。如果错误地设置了 BDR role，TiCDC 同步数据期间无法保证数据正确性和一致性。**
```

> - 一般情况下，禁止在同步的表中使用 `AUTO_INCREMENT` 或 `AUTO_RANDOM` 键，以免产生数据冲突的问题。如果需要使用 `AUTO_INCREMENT` 或 `AUTO_RANDOM` 键，可以通过在不同的集群设置 `auto_increment_increment` 和 `auto_increment_offset` 来使得不同的集群都能够分配到不同的 primary key。假设有三台 TiDB（A、B、C）处于双向同步中，那么你可以采取如下设置：

Reviewer 3: ```suggestion
- 一般情况下，禁止在同步的表中使用 `AUTO_INCREMENT` 或 `AUTO_RANDOM` 键，以免产生数据冲突的问题。如果需要使用 `AUTO_INCREMENT` 或 `AUTO_RANDOM` 键，可以通过在不同的集群设置 `auto_increment_increment` 和 `auto_increment_offset` 来使得不同的集群都能够分配到不同的 primary key。假设有三个 TiDB 集群（A、B、C）处于双向同步中，那么你可以采取如下设置：
```

这样，以上搭建好的集群即可对数据进行双向复制。

## DDL 类别

从 v7.6.0 开始，为了在双向复制中尽可能地支持 DDL 同步，根据 DDL 对业务的影响，TiDB 将 [TiCDC 原本支持同步的 DDL](/ticdc/ticdc-ddl.md) 划分为两种 DDL：可复制的 DDL 和不可复制的 DDL。

### 可复制的 DDL

可复制的 DDL 是指在双向复制中，可以直接执行并同步到其他 TiDB 集群的 DDL。

可复制的 DDL 包括：

- `CREATE DATABASE`
- `CREATE TABLE`

> - `ADD COLUMN`：添加的列必须可以为 `null`或者是同时带有 `not null` 和 `default value` 的列

Reviewer: ```suggestion
- `ADD COLUMN`：添加的列必须可以为 `null`，或者是同时带有 `not null` 和 `default value` 的列
```

- `ADD NON-UNIQUE INDEX`
- `DROP INDEX`
- `MODIFY COLUMN`：仅能修改列的 `default value` 和 `comment`
- `ALTER COLUMN DEFAULT VALUE`
- `MODIFY TABLE COMMENT`
- `RENAME INDEX`
- `ADD TABLE PARTITION`
- `DROP PRIMARY KEY`
- `ALTER TABLE INDEX VISIBILITY`
- `ALTER TABLE TTL`
- `ALTER TABLE REMOVE TTL`
- `CREATE VIEW`
- `DROP VIEW`

### 不可复制的 DDL

不可复制的 DDL 是指对业务影响较大、可能会造成集群间数据不一致性的 DDL，这类 DDL 不能在双向复制中直接通过 TiCDC 同步到其他 TiDB 集群的 DDL。不可复制的 DDL 必须通过特定的操作来执行。

不可复制的 DDL 包括：

- `DROP DATABASE`
- `DROP TABLE`

> - `ADD COLUMN`：添加的列为 `not null` 且不带有 `default value` 的列

Reviewer: ```suggestion
- `ADD COLUMN`：添加的列为 `not null` 且不带有 `default value`
```

- `DROP COLUMN`
- `ADD UNIQUE INDEX`
- `TRUNCATE TABLE`
- `MODIFY COLUMN`：修改列除 `default value` 和 `comment` 以外的属性
- `RENAME TABLE`
- `DROP PARTITION`
- `TRUNCATE PARTITION`
- `ALTER TABLE CHARACTER SET`
- `ALTER DATABASE CHARACTER SET`
- `RECOVER TABLE`
- `ADD PRIMARY KEY`
- `REBASE AUTO ID`
- `EXCHANGE PARTITION`
- `REORGANIZE PARTITION`

## DDL 同步

为了能够解决上述可复制的 DDL 和不可复制的 DDL 两类 DDL 的同步问题，TiDB 引入了两种 BDR role：

- `PRIMARY`：你可以执行可复制的 DDL，但不能执行不可复制的 DDL，可复制的 DDL 会被 TiCDC 同步到下游。

> - `SECONDARY`：你不能执行可复制的 DDL，也不能执行不可复制的 DDL，但是会执行从 TiCDC 同步过来的 DDL。

Reviewer: ```suggestion
- `SECONDARY`：你不能执行可复制的 DDL，也不能执行不可复制的 DDL，但是可以执行从 TiCDC 同步过来的 DDL。
```

> 在不设置 BDR role 时：你可以执行任意 DDL，但是在 TiCDC 开启 `bdr_mode=true` 之后，执行的 DDL 不会被 TiCDC 同步。

Reviewer: ```suggestion
在不设置 BDR role 时，你可以执行任意 DDL。但是在 TiCDC 开启 `bdr_mode=true` 之后，执行的 DDL 不会被 TiCDC 同步。
```

> **警告：**
>
> 双向复制的 DDL 同步目前为实验特性，不建议在生产环境中使用。该功能可能会在未事先通知的情况下发生变化或删除。如果发现 bug，请在 GitHub 上提 [issue](https://github.com/pingcap/tidb/issues) 反馈。

### 可复制的 DDL 的同步场景

1. 选择一个 TiDB 集群，执行 `ADMIN SET BDR ROLE PRIMARY` 将其设置为主集群。
2. 在其他 TiDB 集群上，执行 `ADMIN SET BDR ROLE SECONDARY` 将其设置为从集群。
3. 在主集群上执行**可复制的 DDL**，执行成功的 DDL 会被 TiCDC 同步到从集群中。

> **注意：**
>
> 为了防止误操作：
>
> - 如果在主集群中尝试执行**不可复制的 DDL**，会[报错 8263](/error-codes.md)。
> - 无论在从集群中尝试执行**可复制的 DDL** 还是**不可复制的 DDL**，都会[报错 8263](/error-codes.md)。

### 不可复制的 DDL 的同步场景

1. 对所有 TiDB 集群执行 `ADMIN UNSET BDR ROLE`，撤销集群的 BDR role。
2. 暂停所有集群中需要执行 DDL 的对应的表的写入操作。
3. 等待所有集群中对应表的所有写入已经同步到其他集群后，手动在每一个 TiDB 集群上单独执行所有的 DDL。
4. 等待 DDL 完成之后，重新恢复写入。
5. 按照[可复制的 DDL 的同步场景](#可复制的-ddl-的同步场景)的操作步骤，切换回可复制的 DDL 的同步场景。

> **注意：**
>
> 对所有 TiDB 集群执行 `ADMIN UNSET BDR ROLE` 之后，所有 DDL 都不会被 TiCDC 同步，需要手动在各个集群上分别执行 DDL。

## 停止双向复制

在业务数据停止写入之后，你可以在两个集群中都插入一行特殊的值，通过检查这两行特殊的值来确保数据达到了一致的状态。

检查完毕之后，停止同步任务，并对所有集群执行 `ADMIN UNSET BDR ROLE`。

## 使用限制

- BDR role 只能在以下两种场景中正常使用：

    - 1 个 `PRIMARY` 集群和 n 个 `SECONDARY` 集群（可复制的 DDL 的同步场景）
    - n 个不设置 BDR role 的集群（用于在每个集群手动执行不可复制的 DDL 的同步场景）

>     **注意，请勿将 BDR role 设置为其他情况，例如，既存在集群设置了 `PRIMARY`、`SECONDARY`，又存在集群没有设置 BDR role。如果错误地设置了 BDR role，TiCDC 同步数据期间无法保证数据正确性和一致性。**

Reviewer: ```suggestion
    **注意，请勿将 BDR role 设置为其他情况，例如，既存在设置了 `PRIMARY`、`SECONDARY` 的集群，又存在没有设置 BDR role 的集群。如果错误地设置了 BDR role，TiCDC 同步数据期间无法保证数据正确性和一致性。**
```

> - 一般情况下，禁止在同步的表中使用 `AUTO_INCREMENT` 或 `AUTO_RANDOM` 键，以免产生数据冲突的问题。如果需要使用 `AUTO_INCREMENT` 或 `AUTO_RANDOM` 键，可以通过在不同的集群设置 `auto_increment_increment` 和 `auto_increment_offset` 来使得不同的集群都能够分配到不同的 primary key。假设有三个 TiDB 集群（A、B、C）处于双向同步中，那么你可以采取如下设置：

Reviewer: ```suggestion
- 一般情况下，禁止在同步的表中使用 `AUTO_INCREMENT` 或 `AUTO_RANDOM` 键，以免产生数据冲突的问题。如果需要使用 `AUTO_INCREMENT` 或 `AUTO_RANDOM` 键，可以通过在不同的集群设置 `auto_increment_increment` 和 `auto_increment_offset` 来使得不同的集群都能够分配到不同的主键。假设有三个 TiDB 集群（A、B、C）处于双向同步中，那么你可以采取如下设置：
```

    - 在 A 中设置 `auto_increment_increment=3`，`auto_increment_offset=2000`
    - 在 B 中设置 `auto_increment_increment=3`，`auto_increment_offset=2001`
    - 在 C 中设置 `auto_increment_increment=3`，`auto_increment_offset=2002`

>   这样的话，A、B、C 隐式分配到的 `AUTO_INCREMENT` ID 和 `AUTO_RANDOM` ID 就不会互相冲突。如果需要增加 BDR 模式的集群，则要临时暂停相关业务的写入，重新在所有集群上设置合适的 `auto_increment_increment` 和 `auto_increment_offset` 后，再开启相关业务。

Reviewer: ```suggestion
  这样 A、B、C 隐式分配到的 `AUTO_INCREMENT` ID 和 `AUTO_RANDOM` ID 就不会互相冲突。如果需要增加 BDR 模式的集群，需要临时暂停相关业务的写入，重新在所有集群上设置合适的 `auto_increment_increment` 和 `auto_increment_offset`，然后再开启相关业务。
```

## 使用限制

> - DDL 的限制见[DDL 同步 小节](#ddl-同步)。

Reviewer 2: ```suggestion
- DDL 的限制见[DDL 同步小节](#ddl-同步)。
```

> - 禁止在同步的表中使用 Auto increment/Auto random 键。

Reviewer 3: ```suggestion
- 禁止在同步的表中使用 Auto increment/Auto random 键，否则会产生数据不一致的问题。
```
